pkg/pdp/service/roots_add.go

```diff
@@ -282,3 +338,122 @@ func (p *PDPService) AddRoots(ctx context.Context, id uint64, request []types.Ro
 	}
 	return txHash, nil
 }
+
+// Review: Code below is a copy of github.com/filecoin-project/go-commp-utils/nonffi with support for PieceCIDv2
+// mega unsure
+
+type stackFrame struct {
+	size  uint64
+	commP []byte
+}
+
+func GenerateUnsealedCID(proofType abi.RegisteredSealProof, pieceInfos []abi.PieceInfo) (cid.Cid, error) {
+	spi, found := abi.SealProofInfos[proofType]
+	if !found {
+		return cid.Undef, fmt.Errorf("unknown seal proof type %d", proofType)
+	}
+	if len(pieceInfos) == 0 {
+		return cid.Undef, errors.New("no pieces provided")
+	}
+
+	maxSize := uint64(spi.SectorSize)
+
+	todo := make([]stackFrame, len(pieceInfos))
+
+	// sancheck everything
+	for i, p := range pieceInfos {
+		if p.Size < 128 {
+			return cid.Undef, fmt.Errorf("invalid Size of PieceInfo %d: value %d is too small", i, p.Size)
+		}
+		if uint64(p.Size) > maxSize {
+			return cid.Undef, fmt.Errorf("invalid Size of PieceInfo %d: value %d is larger than sector size of SealProofType %d", i, p.Size, proofType)
+		}
+		if bits.OnesCount64(uint64(p.Size)) != 1 {
+			return cid.Undef, fmt.Errorf("invalid Size of PieceInfo %d: value %d is not a power of 2", i, p.Size)
+		}
+
+		cp, _, err := commcid.PieceCidV2ToDataCommitment(p.PieceCID)
```

Comment:
Main different between this impl and github.com/filecoin-project/go-commp-utils/nonffi